File: 1_SOLID_Design_Patterns/1.01_Single_Responsibility_Principle/Java_Snippet/single_responsibility_principle.java
// Create a helper that can 
// 1. Drive
// 2. Water the plants
// 3. Clean the house
// 4. cook delicious food 

// Violates Single Responsibility Principle
class Helper {
    public void helper() {
        System.out.println("Helper Initialised");
    }

    public String toString() {
        return "I am helper, I can drive, cook, clean and take care of plants";
    }

    public void drive() {
        System.out.println("I am  Driving");
    }

    public void clean() {
        System.out.println("I am  Cleaning");
    }

    public void cook() {
        System.out.println("I am  Cooking");
    }

    public void water_the_plants() {
        System.out.println("I am  watering the plants");
    }
}

// Complies with Single Responsibility Principle
class Driver {
    public void drive() {
        System.out.println("I am  Driving");
    }
}

class Cleaner {
    public void clean() {
        System.out.println("I am  Cleaning");
    }
}

class Chef {
    public void cook() {
        System.out.println("I am  Cooking");
    }
}

class Gardener {
    public void water_the_plants() {
        System.out.println("I am  watering the plants");
    }
}

File: 1_SOLID_Design_Patterns/1.01_Single_Responsibility_Principle/Python_Snippet/single_responsibility_principle.py
'''
Create a helper that can 
1. Drive
2. Water the plants
3. Clean the house
4. cook delicious food 
'''

# Violates Single Responsibility Principle

class Helper(object):
    def __init__(self):
        print("Helper Initialised")

    def __str__(self):
        return f'I am helper, I can drive, cook, clean and take care of plants'

    def drive(self):
        print("I am  Driving")
    
    def clean(self):
        print("I am Cleaning")
    
    def cook(self):
        print("I am Cooking")
    
    def water_the_plants(self):
        print("I am watering the plants")

# Complies with Single Responsibility Principle

class Driver(object):
    def drive(self):
        print("I am  Driving")

class Cleaner(object):
    def clean(self):
        print("I am Cleaning")

class Chef(object):
    def cook(self):
        print("I am Cooking")

class Gardener(object):
    def water_the_plants(self):
        print("I am watering the plants")
